File: python/aho_corasick_string_search/aho_corasick_trie.py
class AhoCorasickNode():
  def __init__(self, letter):
    self.letter = letter
    self.children = []
    self.word = None
    self.prevLink = None
    self.letterChain = None
    self.suffixLink = None
    self.outputLink = None

  def setWord(self, word):
    self.word = word

  def childHasLetter(self, letter):
    found = None
    for child in self.children:
      if child.letter is letter:
        found = child
    return found

  def attachChild(self, child):
    self.children.append(child)

  def display(self, depth = 0):
    for i in xrange(depth):
      print '----|',
    print self.letter or "(root)",
    if self.word:
      print "(", self.word, ")",
    if self.suffixLink:
      print "[", self.suffixLink.letter, '/', self.suffixLink.letterChain, ']',
    if self.outputLink:
      print "{", self.outputLink.word, '}',
    print

    for child in self.children:
      child.display(depth + 1)

class AhoCorasickTrie():
  def __init__(self):
    self.head = AhoCorasickNode('*')
    self.head.letterChain = '*'

  def add(self, word):
    current = self.head
    letterIndex = 0
    while letterIndex < len(word):
      letter = word[letterIndex]
      childWithLetter = current.childHasLetter(letter)

      if childWithLetter:
        current = childWithLetter
      else:
        newNode = AhoCorasickNode(letter)
        newNode.prevLink = current
        newNode.letterChain = word[0:letterIndex+1]
        current.attachChild(newNode)
        current = newNode

      letterIndex += 1

      # final word output
      if letterIndex is len(word):
        current.word = word

    self.buildSuffixes()
    self.buildOutputs()

  def buildSuffixes(self):
    que = [self.head]

    while len(que) > 0:
      item = que.pop(0)

      myPrev = item.prevLink
      if myPrev is self.head:
        item.suffixLink = self.head
      elif myPrev:
        backLink = myPrev
        while backLink.suffixLink:
          backLink = backLink.suffixLink
          backForLink = backLink.childHasLetter(item.letter)
          if backForLink:
            item.suffixLink = backForLink
            break
        if not item.suffixLink:
          item.suffixLink = self.head

      for child in item.children:
        que.append(child)

  def buildOutputs(self):
    que = [self.head]

    while len(que) > 0:
      item = que.pop(0)

      if item.suffixLink:
        if item.suffixLink.word:
          item.outputLink = item.suffixLink
        elif item.suffixLink.outputLink:
          item.outputLink = item.suffixLink.outputLink

      for child in item.children:
        que.append(child)

  def search(self, word):
    outputs = []

    wordIndex = 0
    searchCurrent = self.head
    while wordIndex < len(word):
      letter = word[wordIndex]
      childWithLetter = searchCurrent.childHasLetter(letter)
      while not childWithLetter and searchCurrent is not self.head:
        searchCurrent = searchCurrent.suffixLink
        childWithLetter = searchCurrent.childHasLetter(letter)

      if childWithLetter:
        if childWithLetter.word:
          outputs.append((wordIndex, childWithLetter.word))

        outputLinkChain = childWithLetter.outputLink
        while outputLinkChain:
          outputs.append((wordIndex, outputLinkChain.word))
          outputLinkChain = outputLinkChain.outputLink

        searchCurrent = childWithLetter
      wordIndex += 1

    return outputs

  def display(self):
    self.head.display()

File: ruby/aho_corasick_string_search/aho_corasick_trie.rb
require './aho_corasick_node.rb'

class AhoCorasickTrie

  def initialize
    @head = AhoCorasickNode.new('*')
    @head.letters_sum = '*'
  end

  def add_word(word)
    word_size = word.size
    walking_node = @head
    letter_index = 0

    while letter_index < word_size
      letter = word[letter_index]
      child_with_letter = walking_node.child_of_letter(letter)

      if child_with_letter
        walking_node = child_of_letter
      else
        new_node = AhoCorasickNode.new(letter)
        new_node.prev_link = walking_node
        new_node.letters_sum = word[0..letter_index]
        walking_node.add_child(new_node)
        walking_node = new_node
      end


      letter_index += 1

      if letter_index == word_size
        walking_node.word = word
      end
    end

    self.build_suffixes
    self.build_outputs
  end

  def build_suffixes
    queue = [@head]

    while queue.size > 0
      item = queue.shift

      item_prev = item.prev_link
      if item_prev == @head
        item.suffix_link = @head
      elsif item_prev
        item_back = item_prev
        while item_back.suffix_link
          item_back = item_back.suffix_link
          item_back_child_of_letter = item_back.child_of_letter(item.letter)
          if item_back_child_of_letter
            item.suffix_link = item_back_child_of_letter
            break
          end
        end
        if item.suffix_link == nil
          item.suffix_link = @head
        end
      end

      item.children.each { |c|
        queue << c
      }
    end
  end

  def build_outputs
    queue = [@head]

    while queue.size > 0
      item = queue.shift
      item_suffix = item.suffix_link

      if item_suffix
        if item_suffix.word
          item.output_link = item_suffix
        elsif item_suffix.output_link
          item.output_link = item_suffix.output_link
        end
      end

      item.children.each { |c|
        queue << c
      }
    end
  end

  def search(word)
    outputs = []

    walking_node = @head
    letter_index = 0
    word_size = word.size
    while letter_index < word_size
      letter = word[letter_index]
      child_with_letter = walking_node.child_of_letter(letter)

      while child_with_letter == nil && walking_node != @head
        walking_node = walking_node.suffix_link
        child_with_letter = walking_node.child_of_letter(letter)
      end

      if child_with_letter
        if child_with_letter.word
          outputs << [letter_index, child_with_letter.word]
        end

        output_link_chain = child_with_letter.output_link
        while output_link_chain
          outputs << [letter_index, output_link_chain.word]
          output_link_chain = output_link_chain.output_link
        end

        walking_node = child_with_letter
      end

      letter_index += 1
    end

    return outputs
  end

  def display
    @head.display
  end

end
